pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,2003 @@
+        return cusparsehandle_.get();
+    }
+
+    /**
+     * @brief Utility method for expectation value calculations.
+     *
+     * @param obsName String label for observable. If already exists, will used
+     * cached device value. If not, `gate_matrix` is expected, and will
+     * automatically cache for future reuse.
+     * @param wires Target wires for expectation value.
+     * @param params Parameters for a parametric gate.
+     * @param gate_matrix Optional matrix for observable. Caches for future use
+     * if does not exist.
+     * @return auto Expectation value.
+     */
+    auto expval(const std::string &obsName, const std::vector<size_t> &wires,
+                const std::vector<Precision> &params = {0.0},
+                const std::vector<CFP_t> &gate_matrix = {}) {
+        auto &&par = (params.empty()) ? std::vector<Precision>{0.0} : params;
+        auto &&local_wires =
+            (gate_matrix.empty())
+                ? wires
+                : std::vector<size_t>{
+                      wires.rbegin(),
+                      wires.rend()}; // ensure wire indexing correctly preserved
+                                     // for tensor-observables
+
+        if (!(gate_cache_.gateExists(obsName, par[0]) || gate_matrix.empty())) {
+            gate_cache_.add_gate(obsName, par[0], gate_matrix);
+        } else if (!gate_cache_.gateExists(obsName, par[0]) &&
+                   gate_matrix.empty()) {
+            std::string message =
+                "Currently unsupported observable: " + obsName;
+            throw LightningException(message.c_str());
+        }
+        auto expect_val = getExpectationValueDeviceMatrix(
+            gate_cache_.get_gate_device_ptr(obsName, par[0]), local_wires);
+        return expect_val;
+    }
+    /**
+     * @brief See `expval(const std::string &obsName, const std::vector<size_t>
+     &wires, const std::vector<Precision> &params = {0.0}, const
+     std::vector<CFP_t> &gate_matrix = {})`
+     */
+    auto expval(const std::string &obsName, const std::vector<size_t> &wires,
+                const std::vector<Precision> &params = {0.0},
+                const std::vector<std::complex<Precision>> &gate_matrix = {}) {
+        auto &&par = (params.empty()) ? std::vector<Precision>{0.0} : params;
+
+        std::vector<CFP_t> matrix_cu(gate_matrix.size());
+        if (!(gate_cache_.gateExists(obsName, par[0]) || gate_matrix.empty())) {
+            for (std::size_t i = 0; i < gate_matrix.size(); i++) {
+                matrix_cu[i] = cuUtil::complexToCu<std::complex<Precision>>(
+                    gate_matrix[i]);
+            }
+            gate_cache_.add_gate(obsName, par[0], matrix_cu);
+        } else if (!gate_cache_.gateExists(obsName, par[0]) &&
+                   gate_matrix.empty()) {
+            std::string message =
+                "Currently unsupported observable: " + obsName;
+            throw LightningException(message.c_str());
+        }
```

Comment:
Won't this be taken care of by the next call to `expval`?

Reply:
It is aligned with the corresponding method in statevectorcudamanaged. Marked and will create a PR to change both.